x/delegation/keeper/abci.go

```diff
@@ -24,58 +20,80 @@ func (k Keeper) EndBlock(ctx sdk.Context, _ abci.RequestEndBlock) []abci.Validat
 
-		// get operator slashed proportion to calculate the actual canUndelegated asset amount
-		proportion := k.slashKeeper.OperatorAssetSlashedProportion(ctx, operatorAccAddress, record.AssetID, record.BlockNumber, record.CompleteBlockNumber)
-		if proportion.IsNil() || proportion.IsNegative() || proportion.GT(sdkmath.LegacyNewDec(1)) {
-			panic(fmt.Sprintf("the proportion is invalid,it is:%v", proportion))
+		// calculate the actual canUndelegated asset amount
+		delegationInfo, err := k.GetSingleDelegationInfo(ctx, record.StakerID, record.AssetID, record.OperatorAddr)
+		if err != nil {
+			panic(err)
+		}
+		if record.Amount.GT(delegationInfo.UndelegatableAfterSlash) {
+			record.ActualCompletedAmount = delegationInfo.UndelegatableAfterSlash
+		} else {
+			record.ActualCompletedAmount = record.Amount
 		}
-		canUndelegateProportion := sdkmath.LegacyNewDec(1).Sub(proportion)
-		actualCanUndelegateAmount := canUndelegateProportion.MulInt(record.Amount).TruncateInt()
-		record.ActualCompletedAmount = actualCanUndelegateAmount
 		recordAmountNeg := record.Amount.Neg()
 
 		// update delegation state
 		delegatorAndAmount := make(map[string]*delegationtype.DelegationAmounts)
 		delegatorAndAmount[record.OperatorAddr] = &delegationtype.DelegationAmounts{
-			WaitUndelegationAmount: recordAmountNeg,
+			WaitUndelegationAmount:  recordAmountNeg,
+			UndelegatableAfterSlash: record.ActualCompletedAmount.Neg(),
 		}
 		err = k.UpdateDelegationState(ctx, record.StakerID, record.AssetID, delegatorAndAmount)
 		if err != nil {
 			panic(err)
 		}
 
-		// todo: if use recordAmount as an input parameter, the delegation total amount won't need to be subtracted when the related operator is slashed.
-		err = k.UpdateStakerDelegationTotalAmount(ctx, record.StakerID, record.AssetID, recordAmountNeg)
-		if err != nil {
-			panic(err)
-		}
-
 		// update the staker state
-		err := k.restakingStateKeeper.UpdateStakerAssetState(ctx, record.StakerID, record.AssetID, types.StakerSingleAssetOrChangeInfo{
-			CanWithdrawAmountOrWantChangeValue:      actualCanUndelegateAmount,
-			WaitUndelegationAmountOrWantChangeValue: recordAmountNeg,
+		err = k.assetsKeeper.UpdateStakerAssetState(ctx, record.StakerID, record.AssetID, types.StakerSingleAssetChangeInfo{
+			WithdrawableAmount:  record.ActualCompletedAmount,
+			WaitUnbondingAmount: recordAmountNeg,
 		})
 		if err != nil {
 			panic(err)
 		}
 
 		// update the operator state
-		err = k.restakingStateKeeper.UpdateOperatorAssetState(ctx, operatorAccAddress, record.AssetID, types.OperatorSingleAssetOrChangeInfo{
-			TotalAmountOrWantChangeValue:            actualCanUndelegateAmount.Neg(),
-			WaitUndelegationAmountOrWantChangeValue: recordAmountNeg,
+		err = k.assetsKeeper.UpdateOperatorAssetState(ctx, operatorAccAddress, record.AssetID, types.OperatorSingleAssetChangeInfo{
+			WaitUnbondingAmount: recordAmountNeg,
 		})
 		if err != nil {
 			panic(err)
```

Comment:
Two points here:
- Please remove the `panic`s and use a cached context instead. If there is no error, the cached context's `writeFunc` should be called.
- Add a `TODO` here to say that the message should be sent to the client chain to release the assets for withdrawal. I spoke to @adu-web3 who said that the funds are made available for withdrawal after the undelegation is processed by Exocore. However, that is not the correct approach since the assets are locked for a 10-block (or larger) period during which they can be slashed.

Reply:
> Please remove the panics and use a cached context instead. If there is no error, the cached context's writeFunc should be called.

The handle of errors in `BeginBlock` and `EndBlock` definitely needs to be reconsidered.  What's the meaning of using a cached context? This can be changed in the future PR.

> Add a TODO here to say that the message should be sent to the client chain to release the assets for withdrawal. I spoke to @adu-web3 who said that the funds are made available for withdrawal after the undelegation is processed by Exocore. However, that is not the correct approach since the assets are locked for a 10-block (or larger) period during which they can be slashed.

I remember the previously designed approach was that after undelegation is completed, the user's withdrawable status will increase. Exocore does not need to actively send notifications to the client chain about the completion of undelegation. The client chain will periodically or when the user initiates a transaction, synchronize the withdrawable status from Exocore to the client chain. Additionally, the code here is executed when the unbonding time expires, which means the related assets won't be slashed after this block.

Reply:
For the first point: understood, we will leave this for a future PR. A cached context allows for error handling within the SDK while not modifying the state.
```go
cachedCtx, writeCache := ctx.CacheContext()
if err := k.FunctionThatMightError(ctx); err != nil {
   return err
}
// dump events
ctx.EventManager().EmitEvents(cachedCtx.EventManager().Events())
// store the state change from the function
writeCache()
```

For the second point: The current implementation, in my understanding of the smart contract, is that whenever an undelegation request is executed on the client chain, a message reaches the Exocore chain. The message is executed in the `ExocoreGateway.sol`, which calls the undelegation precompile. The call's result (success or failure) is sent back to the client chain gateway. The client chain gateway then allows withdrawal of the undelegated assets, which is an incorrect implementation. The undelegated assets should remain stuck for 10 blocks (or the unbonding period, whichever is higher).

To fix this, my approach is to remove the immediate release feature from the client chain gateway, and for Exocore chain to send a message to the client chain when the assets are available. At this point, we may have to consider batching such messages. This is the TODO I was asking you to add here; however, it is not a blocker for this PR.

Reply:
When the undelegation call's result is sent back to the client chain, It's just a message to notify the client chain that the undelegation has been successfully routed and queued in the Exocore chain. It doesn't mean the asset related to this undelegation can be withdrawn because the withdrawable amount increases only when this undelegation is completed.  As for how to notify the client chain after completing the undelegation, the client chain contract will sync the withdrawable state from the Exocore chain regularly or when a staker initiates actions(such as deposit, delegation, undelegation and withdrawing) from the client chain. If you find the withdrawable amount will increase immediately when
the undelegation call's result is successfully sent back to the client chain, there might be a bug in the client chain contract. Please let me know about it.

Reply:
Appreciate for providing the cache context code.

Reply:
I misunderstood the Solidity implementation; please ignore the second part of my original comment.